crates/c-api/src/func.rs

```diff
@@ -187,25 +187,38 @@ pub struct wasmtime_caller_t<'a> {
     caller: Caller<'a, crate::StoreData>,
 }
 
+pub type wasmtime_func_callback_t = extern "C" fn(
+    *mut c_void,
+    *mut wasmtime_caller_t,
+    *const wasmtime_val_t,
+    usize,
+    *mut wasmtime_val_t,
+    usize,
+) -> Option<Box<wasm_trap_t>>;
+
 #[no_mangle]
 pub unsafe extern "C" fn wasmtime_func_new(
     store: CStoreContextMut<'_>,
     ty: &wasm_functype_t,
-    callback: extern "C" fn(
-        *mut c_void,
-        *mut wasmtime_caller_t,
-        *const wasmtime_val_t,
-        usize,
-        *mut wasmtime_val_t,
-        usize,
-    ) -> Option<Box<wasm_trap_t>>,
+    callback: wasmtime_func_callback_t,
     data: *mut c_void,
     finalizer: Option<extern "C" fn(*mut std::ffi::c_void)>,
     func: &mut Func,
 ) {
-    let foreign = crate::ForeignData { data, finalizer };
     let ty = ty.ty().ty.clone();
-    let f = Func::new(store, ty, move |caller, params, results| {
+    let cb = call_callback(callback, data, finalizer);
+    let f = Func::new(store, ty, cb);
+    *func = f;
+}
+
+#[no_mangle]
+pub(crate) unsafe fn call_callback(
```

Comment:
I might have named this `callback_to_fn`, that was the only part of reading this that tripped me up.

Reply:
Good point, let's go with `c_callback_to_rust_fn`!